bridge/src/reactor.rs

```diff
@@ -0,0 +1,108 @@
+use crate::connection::Client;
+use temporal_sdk_core::api::{errors, Worker};
+use temporal_sdk_core_protos::coresdk::activity_task::ActivityTask;
+use thiserror::Error;
+use tokio::runtime::{Runtime};
+use tokio::sync::{mpsc, oneshot};
+use std::sync::Arc;
+
+pub enum Request {
+    InitWorker {
+      config: temporal_sdk_core::WorkerConfig,
+      client: Client
+    },
+    PollActivityTask {
+        worker: Arc<temporal_sdk_core::Worker>
+    },
+}
+
+pub enum Response {
+    Worker {
+      worker: temporal_sdk_core::Worker
+    },
+    ActivityTask {
+      task: ActivityTask
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum ReactorError {
+    #[error(transparent)]
+    UnableToPollActivityTask(errors::PollActivityError),
+}
+
+impl std::fmt::Debug for Response {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "We have a response")
+    }
+}
+
+pub type ReactorResult = Result<Response, ReactorError>;
+
+/// A request wrapper which includes a channel for sending back the response
+struct Envelope {
+    request: Request,
+    channel: oneshot::Sender<ReactorResult>
+}
+
+impl std::fmt::Debug for Envelope {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "We have an envelope")
+    }
+}
+
+async fn handle_request(request: Request, channel: oneshot::Sender<ReactorResult>) {
+    match request {
+        Request::InitWorker { config, client } => {
+            let worker = temporal_sdk_core::init_worker(config, client.into_inner());
+
+            let response = Response::Worker { worker };
+
+            channel.send(Ok(response)).expect("Reactor was unable to send a response via channel");
+        },
+        Request::PollActivityTask { worker } => {
+            let result = (&*worker).poll_activity_task().await
+                .map(|task| Response::ActivityTask { task })
+                .map_err(|e| ReactorError::UnableToPollActivityTask(e));
+
+            channel.send(result).expect("Reactor was unable to send a response via channel");
+        },
+    }
+}
+
+fn start_loop(runtime: Arc<Runtime>, mut receiver: mpsc::UnboundedReceiver<Envelope>) {
+    runtime.block_on(async move {
+        while let Some(msg) = receiver.recv().await {
+            tokio::spawn(handle_request(msg.request, msg.channel));
+        }
+    });
+}
+
+/// Responsible for processing requests asynchronously using an event loop
+/// while exposing a synchronous interface to the callers
+pub struct Reactor {
```

Comment:
I thought a bit more about this and it might actually make sense and will reduce the complexity a bit. I wonder if that has the same performance characteristics though. Right now we have a separate thread listening on a channel and processing requests in new tokio tasks.

You're suggesting that just calling `runtime.block_on` performs similarly even when called at the same time across multiple threads, right?

Reply:
Perf is irrelevant here so I wouldn't base any decisions on that. The limiting factor here will always be the actual network rather than any of these internals.

Block on is probably a bit easier to read than setting up the oneshot and immediately blocking on it, so I think it makes some sense to do. The `process` function should just be clear in its docstring that it blocks the thread, possibly for a long time, when called.